test/Gen.hs

```diff
@@ -130,15 +147,68 @@ w = genT
 
 type W = T "W"
 
+unit :: Gen ()
+unit = atom "()" ()
+
+identity :: Gen (a -> Identity a)
+identity = atom "Identity" Identity
+
 fn :: (Fn.Arg a, Fn.Vary a, Show a) => Gen b -> Gen (a -> b)
 fn b = Gen (lift (fmap (fmap runTerm) . showingFn <$> Fn.fn (fst <$> runWriterT (runGen b))))
 
+termFn :: Gen b -> Gen (a -> b)
+termFn b = Gen $ recursive Hedgehog.Gen.choice
+  [ runGen (atom "const" const <*> b) ]
+  []
+
 choice :: [Gen a] -> Gen a
 choice = Gen . Hedgehog.Gen.choice . Prelude.map runGen
 
 integral :: (Integral a, Show a) => Range a -> Gen a
 integral range = Gen (showing <$> Hedgehog.Gen.integral range)
 
+unicode :: Gen Char
+unicode = Gen (showing <$> Hedgehog.Gen.unicode)
+
+string :: Range Int -> Gen Char -> Gen String
+string range cs = Gen (showing <$> Hedgehog.Gen.string range (runTerm <$> runGen cs))
+
+
+newtype Run f g m = Run (forall a . f (m a) -> PureC (g a))
+
+type RunL g m = Run Identity g m
+
+pattern RunL :: (forall a . m a -> PureC (f a)) -> Run Identity f m
+pattern RunL run <- Run ((.# Identity) -> run) where
+  RunL run = Run (run . runIdentity)
+
+{-# COMPLETE RunL #-}
+
+type RunR f m = Run f Identity m
+
+pattern RunR :: (forall a . f (m a) -> PureC a) -> Run f Identity m
+pattern RunR run <- Run ((fmap runIdentity #.) -> run) where
+  RunR run = Run (fmap Identity . run)
+
+{-# COMPLETE RunR #-}
+
+type RunS s m = Run ((,) s) ((,) s) m
+
+pattern RunS :: (forall a . s -> m a -> PureC (s, a)) -> Run ((,) s) ((,) s) m
+pattern RunS run <- Run (curry' -> run) where
+  RunS run = Run (uncurry run)
```

Comment:
This just handles the currying/uncurrying for `StateC`/`FreshC`, cos it was a little ugly doing them manually.